File: algorithms/Max Consecutive Ones III/Max Consecutive Ones III.js
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function(nums, k) {
    let left = 0;
    let oneCount = 0;
    let maxLength = 0;
    
    for (let right = 0; right < nums.length; right ++) {
        if (nums[right]) {  // so if the element is a 1 since 0 is falsy
            oneCount++
        }
        
        if ((right - left + 1 - oneCount) > k) { // check if we've used all of our replacements
            if (nums[left]) { // start shrinking the window if its a 1 we subtract a count from oneCount
                oneCount--
            }
            left++
        }
        
        maxLength = Math.max(maxLength, right - left + 1); // update maxLength each iteration for largest window
    }
    
    return maxLength
};

File: algorithms/M/Max Consecutive Ones III/Max Consecutive Ones III.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        
        j = 0
        
        ans = 0
        
        result = -999999999
        
        for i in range(len(nums)):
            if nums[i]==0:
                ans+=1
            elif ans>k:
                result = max(result,j-i)

            while ans>k:
                if nums[j]==0:
                    ans-=1
                j+=1

        return max(result,i-j+1)